apps/api/src/deployment/services/top-up-custodial-deployments/top-up-custodial-deployments.service.ts

```diff
@@ -0,0 +1,136 @@
+import { AllowanceHttpService, BalanceHttpService, BlockHttpService, DeploymentAllowance } from "@akashnetwork/http-sdk";
+import { singleton } from "tsyringe";
+
+import { InjectSigningClient } from "@src/billing/providers/signing-client.provider";
+import { InjectWallet } from "@src/billing/providers/wallet.provider";
+import { MasterSigningClientService, MasterWalletService } from "@src/billing/services";
+import { LoggerService } from "@src/core";
+import { InjectSentry, Sentry } from "@src/core/providers/sentry.provider";
+import { SentryEventService } from "@src/core/services/sentry-event/sentry-event.service";
+import { DeploymentConfig, InjectDeploymentConfig } from "@src/deployment/config/config.provider";
+import { DrainingDeploymentOutput, LeaseRepository } from "@src/deployment/repositories/lease/lease.repository";
+import { averageBlockCountInAnHour } from "@src/utils/constants";
+
+interface Balances {
+  denom: string;
+  feesLimit: number;
+  deploymentLimit: number;
+  balance: number;
+}
+
+@singleton()
+export class TopUpCustodialDeploymentsService {
+  private readonly CONCURRENCY = 10;
+
+  private readonly MIN_FEES_AVAILABLE = 5000;
+
+  private readonly logger = new LoggerService({ context: TopUpCustodialDeploymentsService.name });
+
+  constructor(
+    private readonly allowanceHttpService: AllowanceHttpService,
+    private readonly balanceHttpService: BalanceHttpService,
+    private readonly blockHttpService: BlockHttpService,
+    @InjectWallet("UAKT_TOP_UP") private readonly uaktMasterWalletService: MasterWalletService,
+    @InjectWallet("USDC_TOP_UP") private readonly usdtMasterWalletService: MasterWalletService,
+    @InjectSigningClient("UAKT_TOP_UP") private readonly uaktMasterSigningClientService: MasterSigningClientService,
+    @InjectSigningClient("USDC_TOP_UP") private readonly usdtMasterSigningClientService: MasterSigningClientService,
+    private readonly leaseRepository: LeaseRepository,
+    @InjectDeploymentConfig() private readonly config: DeploymentConfig,
+    @InjectSentry() private readonly sentry: Sentry,
+    private readonly sentryEventService: SentryEventService
+  ) {}
+
+  async topUpDeployments() {
+    const wallets = [
+      { wallet: this.uaktMasterWalletService, client: this.uaktMasterSigningClientService },
+      { wallet: this.usdtMasterWalletService, client: this.usdtMasterSigningClientService }
+    ];
+
+    const topUpAllCustodialDeployments = wallets.map(async ({ wallet, client }) => {
+      const address = await wallet.getFirstAddress();
+      await this.allowanceHttpService.paginateDeploymentGrants({ grantee: address, limit: this.CONCURRENCY }, async grants => {
```

Comment:
Isn't the master wallet the **granter** in this case?

Reply:
no, custodial wallets are granting master wallet to top up on their behalf